### PR TITLE
feat: Clarifying test types and expectations

### PR DESCRIPTION
**Pull request by craig-landry**
This largely states that we should be writing a lot of unit tests, a decent amount of integration tests, and very few e2e tests.

### CODE REVIEW

**Review by jkuester — COMMENTED** [review]
This is so great!  I have left my suggestions/opinions inline and am eager to continue the conversation more!

File: content/en/contribute/code/core/automated-tests.md [hunk 1]
```diff
@@ -117,3 +126,24 @@ Running e2e tests can be quite slow so to save time modify the `specs` property
...
+1. Unit tests
+1. Integration tests
+1. end-to-end tests (sometimes called "feature" or "system" tests)
+
+The different test types tend to increase in cost, complexity, time to run, and fragility in the order listed above. Below is a table outlining these factors:
+
+
+| Test Type         | Description| Expectations | Execution Speed     | Complexity | Fragility | 
+|--------------|-----------------|---------------------|------------|---------|---------| 
+| Unit  | Small tests of specific behavior. Often written to handle a specific condition within a specific function/component. Any dependencies are often mocked.  | High coverage of functionality. If measured in branch coverage percentage, aim for >= 90%. | Extremely fast | Extremely low | Extremely stable |
```
[the remark this answers is not in the document]
Comment (craig-landry) [reply content/en/contribute/code/core/automated-tests.md:143]: I'm all for increasing this! If we start talking about reducing it, I have more to say, but if this is about how much to *increase*, I'm good for whatever the team likes here. I'm going to leave this suggestion uncommitted for now to make sure others can weigh in.

File: content/en/contribute/code/core/automated-tests.md [hunk 2]
```diff
@@ -117,3 +126,24 @@ Running e2e tests can be quite slow so to save time modify the `specs` property
...
+| End-to-end  | Verify that major system features work as expected. This means things like logging in, submitting a form, showing/hiding a tab based on configuration. Another way of thinking of it is that it's a "sanity check". | Very few. This isn't to test every single thing or even close to it. Something close to 20 to 30 (total!) is reasonable. This is not the place to guarantee confidence in the system. The two types above are for that. | Painfully slow. So slow that too many of them can lead to developers not bothering to run them! | Low for the test itself (click tab, enter text into form, click submit, check text on screen. Extremely high for the setup. | Painful fragility with high risk of race conditions and high maintenance burden. |
```
[the remark this answers is not in the document]
Comment (craig-landry) [reply content/en/contribute/code/core/automated-tests.md:145]: I see you found the meat of the discussion/PR here! 😉 I appreciate the consideration and emphasis on "unless". Our constraints here are slow, fragile tests vs quality currently. I appreciate that we're choosing quality. The challenge is how to go faster and stable with quality (I want both please!).  I think the answer is in your words of (to avoid) "whole CHT setup". 

I very much like the idea of isolating components and then testing them, and it's cool to have new work showing such an approach already! Testing strategy aside, it builds upon the already-good practice of separation of concerns, and the general pattern recommended by web components. As a related example, I like development of front-end components with things like [Storybook](https://storybook.js.org/) where you build that way, in small parts, and then of course can test that way too. It's a nice side effect that comes with such efforts, which is a funny way to put it because often people start with something like Storybook if only for the joy of building UI components that way.

Back to the testing aspects... if I understand correctly here, we're still talking about running tests with a full web browser (slow), but avoiding all the risks in a true "end to end" (fragile and full of dependencies) test. Execution time goes up somewhat thanks to less initialization/setup when running tests, but we still have to wait on the browser startup and click events. There are still risks of race conditions, although far fewer since we've cut out all the other parts (like the DB).

_⚠ Warning, sketchy suggestion follows:_
There may be a route to getting the speed attribute here too. If we get our separation of concerns right (at least at the web component level) we could detect what changed and only execute the UI component tests for those. On average, that would mean running almost zero of them. That wouldn't work for an e2e (or even integration) test since those involve too many dependencies to be guaranteed of when it's safe to skip running a test. If we're doing a great job on isolated components though it does become safe.

File: content/en/contribute/code/core/automated-tests.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply content/en/contribute/code/core/automated-tests.md:143]: "aim for 100%" (for new code) makes sense to me. There will always be situations where it's not possible so it wouldn't be a good absolute rule.

File: content/en/contribute/code/core/automated-tests.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (garethbowen) [reply content/en/contribute/code/core/automated-tests.md:145]: I expect the point of this document to be current best practice, rather than desirable future state. Given that, stating a number of total tests that is an order of magnitude smaller than the current suite probably isn't going to help anyone.

It would probably be more useful to talk about when to add a new e2e test, or some sort of talk about coverage in terms of pages covered or similar.

[the remark this answers is not in the document]
Comment (tatilepizs) [reply content/en/contribute/code/core/automated-tests.md:145]: It is really hard for me to comment on this topic because I don't have a solution. I love the e2e tests because it is what we do when we used to do manual tests, and I have to be honest, I really like to automate things 😅  But I also agree that it is painful to see how much time takes to run them and how fragile they are.

I loved the way the e2e tests were isolated for the web component, they are faster and as effective as the e2e test tests that we have for the cht-core but without having the complete setup. It would be great if we could have something similar for other sections of the system.

About what Gareth was saying related to when to add new e2e tests (that is something Josh also mentioned in his comment) we have discussed this in our QA syncs, how to have a "rule" or a "guide" that the developers can follow every time that they create a new feature. In a meeting that we had, Josh mentioned that he has a "personal rule", and every time that he changes something in the `webapp` it indicates that a new test (or a new scenario) needs to be added because it will be the only way to actually test it, and I think that it is a really good idea, and easy to follow. 
I don't know that @craig-landry or @garethbowen think about that?

Comment (tatilepizs) [inline content/en/contribute/code/core/automated-tests.md:145]:
```suggestion
| End-to-end  | Verify that major system features work as expected. This means things like logging in, submitting a form, showing/hiding a tab based on configuration. Another way of thinking of it is that it's a "sanity check". | Very few. This isn't to test every single thing or even close to it. Something close to 20 to 30 (total!) is reasonable. This is not the place to guarantee confidence in the system. The two types above are for that. | Painfully slow. So slow that too many of them can lead to developers not bothering to run them! | Low for the test itself (click tab, enter text into form, click submit, check text on screen). Extremely high for the setup. | Painful fragility with high risk of race conditions and high maintenance burden. |
```

**Review by tatilepizs — COMMENTED** [review]
This is something that we have talked about for months, thank you @craig-landry for starting the discussion!

File: content/en/contribute/code/core/automated-tests.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (craig-landry) [reply content/en/contribute/code/core/automated-tests.md:145]: > rather than desirable future state

Good point; agreed.

[the remark this answers is not in the document]
Comment (tatilepizs) [reply content/en/contribute/code/core/automated-tests.md:145]: It is really hard for me to comment on this topic because I don’t have a solution. I love the e2e tests because it is what we do when we used to do manual tests, and I have to be honest, I really like to automate things :sweat_smile: But I also agree that it is painful to see how much time takes to run them and how fragile they are.
I loved the way the e2e tests were isolated for the web component, they are faster and as effective as the e2e test tests that we have for the cht-core but without having the complete setup. It would be great if we could have something similar for other sections of the system.

About what Gareth was saying related to when to add new e2e tests (that is something Josh also mentioned in his comment) we have discussed this in our QA syncs, how to have a “rule” or a “guide” that the developers can follow every time that they create a new feature. In a meeting that we had, Josh mentioned that he has a “personal rule”, and every time that he changes something in the webapp it indicates that a new test (or a new scenario) needs to be added because it will be the only way to actually test it, and I think that it is a really good idea, and easy to follow.

I don’t know what @craig-landry or @garethbowen think about that?

File: content/en/contribute/code/core/automated-tests.md [hunk 3]
```diff
@@ -117,3 +126,24 @@ Running e2e tests can be quite slow so to save time modify the `specs` property
...
```
Comment (lorerod) [inline content/en/contribute/code/core/automated-tests.md:145]:
```suggestion
| End-to-end  | Verify that major system features work as expected, it aim to verify the functionality of the application as a whole, mostly from the user interface. This means things like logging in, submitting a form, showing/hiding a tab based on configuration. | Dramatically fewer than unit tests. And we should evaluate first if, with an integration test, you can cover the scenario, then better do it in that layer. For coverage, we should consider feature requirements and evaluate if the requirements already have related tests. | Painfully slow. So slow that too many of them can lead to developers not bothering to run them! | Low for the test itself (click tab, enter text into form, click submit, check text on screen). Extremely high for the setup. | Painful fragility with high risk of race conditions and high maintenance burden. |
```

File: content/en/contribute/code/core/automated-tests.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lorerod) [reply content/en/contribute/code/core/automated-tests.md:145]: This is the controversial part of the document. 
We mainly test UI features in this layer, and I agree with Josh that if we don´t have a way to test web app features on other layers, it will be challenging to reduce the number of e2e tests. E2e tests give us the most confidence to decide if your feature is working; we need to start changing the current setup if we want to change that fact.
We can also make some effort about duplicate scenarios; we need to clean up the scenarios covered on our e2e test and define main workflows that cover multiple features or something similar to try to reduce the current number of tests. This cleanup may also help when deciding whether to add a whole new test or some extra steps to existing ones.

**Review by lorerod — CHANGES_REQUESTED** [review]
Thank you, Craig, for taking the initiative on this.

**Comment by craig-landry** [comment]
Thank you all for the thoughtful remarks on this. As I read this at the moment, it seems like roughly this:
1. E2E tests are slow and fragile. Nobody likes those aspects.
1. More tests generally greater confidence.
1. Clear guidance is necessary for knowing when to add more E2E tests (or not).
1. Some benefit can be had by eliminating duplicates.

I think fourth one can just be done as a good practice of maintaining a clean codebase.

To the first and second items, I think there is a lot to be gained by respecting the difference in UI component tests and E2E tests. As I noted in an earlier comment, I really like how building and testing components that way encourages other best-practices in development.

**Proposal:**
Add a new test type to the table above for UI component tests. The guidance would be that they should be isolated, but the test assertions would be much more thorough than what is called for in E2E tests. This would also mean leaving the current E2E section as is, except for adding clarity on when to write them.

What do you all think?

**Comment by garethbowen** [comment]
> Add a new test type to the table above for "UI component tests".

If I understand this correctly it's a bit like a unit test for the UI, so you can, for example, make sure the datepicker behaves in a wide range of edge cases, different languages, leap years, whatever and the right date ends up in the input field, but what you _don't_ check is any other part of the integration, eg: don't bother submitting the form. This will mean they run much faster and probably more reliably. Is that what you were thinking?

We'll need to figure out _how_ to write these. One solution would be to separate out the UI as Web Components or similar which forces us to write them in a standalone way so they can run without the rest of the CHT chrome. @jkuester is leading the way here with the enketo component and may have some advice...

**Comment by garethbowen** [comment]
@craig-landry Can you please revive this PR?

**Comment by craig-landry** [comment]
> @craig-landry Can you please revive this PR?

Yes, although I won't be able to until March. I'll pick this up then. I apologize for the extended delay!

**Comment by lorerod** [comment]
Hi @craig-landry, can I continue with this?

**Comment by craig-landry** [comment]
Thanks for getting back to this @lorerod. I keep failing to find time to properly get this worked out. I don't like blocking everyone on it though. Yes, please do work with the team on this! Whatever the team feels good about is cool with me. With that, I know this has conflicts and also has text needing changes in line with QA and dev preferences. If it's easier to throw this away and make a new page, that's ok too.

**Comment by andrablaj** [comment]
@lorerod @craig-landry this PR can be closed now, right, as the changes were merged via #1376 ?

**Comment by lorerod** [comment]
Yes, I'm closing this now. Thanks for cleaning.
